### DISCUSSION TITLE
Using caller from within procedures

### DISCUSSION THREAD

**Discussion by eensander** [trpc/trpc#4839] [Q&A]
Hello,

So in the documentation, it states that a caller should not be used to call other procedures from within a procedure since it will lead to a potential overhead. I was wondering how significant this overhead is and if, next to the overhead, there are other unintended consequences. 

The reason for this question is that I think that in my situation, creating a dedicated DL (data layer) for only this use case is not needed. Furthermore, it is nice to have input validation at the same time. I get that the middleware might get executed again, though this is still fine. 

Then, there is the mention of the potential overhead of re-creating the context. Though, this won't be the case when you simply pass the context, would it? Like this;

```ts
'task': t.procedure
  .input(z.number())
  .mutation(async ({ctx, input}) => {
    const caller = appRouter.createCaller(ctx);
    ...
  })
```

So, I can live with some overhead in a few infrequently used procedures, if that means that I can call other procedures without having to write a complete separate DL. Considering this, are there other significant disadvantages? 

Thanks in advance!

**Comment by DawidWraga** [comment]
Yes this would be really helpful for creating modular procedures and then composing them into more complex ones. Creating functions outside of the procedure and then wrapping around them feels clunky.

an API like this would be very helpful
```ts
procedure1: t.procedure
  .mutation(async ({ctx, caller, input}) => {
  })
procedure2 t.procedure
  .mutation(async ({ctx, caller, input}) => {
      caller.procedure1()
  })
```

Not sure if this would be possible without breaking typescript

**Reply by Nick-Lucas** [reply]
Every other API framework in the world recommends using Service classes and making Controllers minimal. tRPC isn't much different in this respect. It's a good idea to separate your business logic from the API logic and makes everything more easily testable and composable.

If you don't want to be too abstracted, then using Context Extension in a nearby middleware is a great solution.

```ts

const myprocedure = t.procedure.use((opts) => 
  opts.next({ ctx: { foo: () => console.log("foo") } })
)

procedure1: myprocedure
  .mutation(async ({ctx, caller, input}) => {
    ctx.foo()
  })
procedure2 myprocedure
  .mutation(async ({ctx, caller, input}) => {
      ctx.foo()
  })
```

using createCaller executes the whole middleware chain though, which is why you wouldn't want to do it within a procedure

**Reply by DawidWraga** [reply]
Before I left my comment, I had a similar idea but I think I didnt go about it in the best way.

I tried to separate my business logic into service classes. I started to create an abstract service class which stores a user ID, db and zod validation... Then I realized that I'm basically trying to recreate much of trpc and decided it was a waste of time.

I don't have that much experience with service/controller architecture so I'm eager and curious to understand how you would implement your suggestion to a trpc workflow.

I like your suggestion about using the context in middlware, thank you I appreciate it.

**Reply by DawidWraga** [reply]
Hey @Nick-Lucas after this discussion I created a service builder inspired by tRPC procedure builder and published it to npm along with some docs + example project which you can see [here ](https://www.davstack.com/service/overview)

I'd really appreciate to hear your thoughts on this package and I'm curious to ask - do you think something this could ever make it into tRPC? by this I mean first class support for defining procedures outside of routers to separate your business logic from the API logic.

Also, I mentioned your comment /name in the docs and wanted to make sure that you're okay with that.

**Reply by eensander** [reply]
@DawidWraga Although your reply was not aimed at me, I'll take the opportunity to share my opinion on it. Please note that I am pretty opinionated. 

The general concept of the package seems very promising to me! I also like the direct integration with tRPC using the `createTrpcRouterFromServices` and `createTrpcProcedureFromService` methods. 

Since I started this thread, I'm obviously also fond of the section [direct service usage](https://github.com/DawidWraga/davstack/tree/main/packages/service#direct-service-usage). Though, it is currently unclear to me if this approach actually solves the problem I described in my original comment: when calling another service from within a service, are the middlewares of the second middlewares evaluated again? Since this could lead to overhead. As I mentioned in my original comment, this is not necessarily a big problem for _me_. Though, I could potentially see it being a problem for larger applications. 

Moreover, I wonder how significant the need for (authorization/authentication) middlewares to exist in the service layer of an application is: I think that is is more the responsibility for the router part (tRPC in this case) to properly authorize users. Seperately, I think the methods should contain all relevant inputs as arguments: instead of getting the user from the context taking it as an extra argument. This leads to more reproducible functions, which I think is a good practice.

So instead of:
```ts
const getTasks = authedService
  .input(z.object({ projectId: z.string() }))
  .query(async ({ ctx, input }) => {
    return ctx.db.tasks.findMany({ where: { userId: ctx.user.id, projectId: input.projectId } });
  });
```

More like:
```ts
const getTasks = service() // not using authorization middlewares that get the user_id
  .input(z.object({ userId: z.string(), projectId: z.string() })) // pass userId as "argument" (aka input)
  .query(async ({ ctx, input }) => {
    return ctx.db.tasks.findMany({ where: { userId: input.userId, projectId: input.projectId } });
  });
```

This is also an opportunity to make the package more lightweight, having one purpose: defining the services. This is how I would see it working together with tRPC:

### Service Layer
```ts
// services/lib.ts:
import { initDavstack } from '@davstack/service';

export function serviceContext() {
  return { db };
}

const davStack = initDavstack(serviceContext())
export const services = davStack.services
export const service = davStack.service
```

```ts
// services/project.service.ts:
import { services, service } from './lib';
import { z } from 'zod';

export const projectServices = services({
  getProjects = service()
    .input(z.object({ userId: z.string().optional(), projectId: z.string().optional() }))
    .output(
      z.object({
        id: z.string(),
        name: z.string(),
      })
    .query(async ({ ctx, input: conditions }) => {
      const projects = await ctx.db.from('projects').select('*').where(conditions)
    }),

  getProjectsOfUser = service()
    .input(z.object({ userId: z.string() }))
    .output(
      z.object({
        id: z.string(),
        name: z.string(),
      })
    )
    .query(async ({ ctx, input }) => {
      const projects = await caller.project.getProjects({userId: ctx.user.id})
    }),
  // ...
})
```

```ts
// services/index.ts:
import { projectServices } from './project.service.ts'

const appServices = services({
  project: projectServices
})
```

### Router Layer

```ts
// trpc/index.ts:

// ... (general setup)

// borrowed from comment above https://github.com/trpc/trpc/discussions/4839#discussioncomment-8224476 :
const serviceProcedure = t.procedure.use((opts) => 
  opts.next({ ctx: { services: appServices } })
)

const authorizedProcedure = serviceProcedure.use(/* authorization logic */)
```

```ts
// trpc/routers/project.router.ts:

// purpose of router's procedure: validate that user is authenticated and choose the right one
export const projectRouter = router({
  'projects': authorizedProcedure
    .query(async ({ ctx }) => ctx.services.project.getProjectsOfUser({userId: ctx.user.id})),
  // ...
})
```

The above code is just an example of what I think syntactically makes sense, in combination with my feedback. Though, I didn't test any of the code so please take it with a spoon of salt. Another consideration with my suggestion is that now you have to define the zod validation rules twice: in the procedure and services. And these can be different, since the service inputs could include the user while the procedure would usually not. Not sure what I think about that at this moment. 
Also note the example I did with calling a service from within a service might not be possible the way it is demonstrated with caller. Might have to get creative here.

Please don't take my suggestion too strongly. I'd like to reiterate that I think the concept and work that you did is really nice and valuable. I'm definitely going to keep an eye on the project and perhaps consider using it later. Cheers!

**Comment by david8z** [comment]
Hello @Nick-Lucas I know this is from long time ago but how would you structure correctly a project using service controller approach with TRPC?

I feel I am repeating myself a lot by trying to do so.
